### PR TITLE
bpo-38314: Add is_reading() method to asyncio _UnixReadPipeTransport

### PR DESCRIPTION
**Pull request by calluw**
Unix read pipe (from `Lib:asyncio.unix_events`) is missing `is_reading()` method available for other read transports within the `asyncio` protocol, which checks that the transport read is not
paused or the transport is not closing. 

This behaviour is documented in `Doc/library/asyncio-protocol.rst`.

Additional tests for the `Lib:asyncio.unix_events` module which check against the state of the pipe and the behaviour of `is_reading()`.

https://bugs.python.org/issue38314

### CODE REVIEW

**Comment by callumquick** [comment]
@aeros Hi Kyle, does this need a news or could you add the skip news label for me? I am contributing as part of the EnHackathon team: https://enhackathon.github.io/.

**Comment by asvetlov** [comment]
Please avoid mocking but write a test that uses a pair of *real* sockets connected via loopback interface (localhost).

Mocking is highly not reliable; we have a lot of problems in asyncio behind of this.

P.S.
Yes, NEWs is required for this feature.

**Review by asvetlov — CHANGES_REQUESTED** [review]
Please fix my comments.
The code change itself is correct.

**Comment by callumquick** [comment]
Hi @asvetlov, thanks for your comments on the first draft. However, I'd like to ask for some clarification on the changes you've asked for:
- Within this test suite, the style of tests is distinct from integration tests like opening sockets into more unit-based tests by mocking the pipe and file descriptors: should we break from this style of tests within the same test suite?
- Also, I am not sure how `is_reading()` would translate in the socket opening integration test asked for when it just checks internal state: its functionality is constrained/verified by the current unit test. 

If the requested changes are needed for merge, would you be able to give some guidance on making them?
This is my first contribution PR so apologies if I have misunderstood something.

**Comment by asvetlov** [comment]
The asyncio test suite is in the middle of translation from primarily mock-based tests to almost solely functional ones (this is our target).

Maybe a separate test case in the same file fits better to our needs. Currently, there are a lot of functional tests inside test_events.py; it makes the test file unmaintainable.

That's why unit test pipes better to sit in `test_unix_events.py` (or even not existing yet `test_unix_pipes.py`).

The test scenario could be:
1. Create a pipe pair.
2. Create a pipe reader (`loop.connect_read_pipe()`)
3. Check `pipe.is_reading()`
4. Call `pipe.pause_reading()`
5. Check `pipe.is_reading()`
6. Call `pipe.resume_reading()`
7. Check `pipe.is_reading()`
8. Call `pipe.close()`
9. Check `pipe.is_reading()`


Something like this.

**Comment by aeros** [comment]
@callumquick 
> Hi Kyle, does this need a news or could you add the skip news label for me?

I think this one could benefit from a brief news entry since it's implementing a new public method (even if the new method is fairly simple). I'll add an asyncio label though.

File: Misc/NEWS.d/next/Library/2019-11-04-17-20-43.bpo-38314.zWz6_P.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+The read pipe on UNIX transports now has access to the `is_reading()` API which is noted in the documentation for asyncio protocols as being generally available on read transports.
```
Comment (aeros) [inline Misc/NEWS.d/next/Library/2019-11-04-17-20-43.bpo-38314.zWz6_P.rst]: I think we could make this more succinct and briefly mention that it's related to asyncio. Also, the max line width of MISC/news entries has to be within 79 chars. 
```suggestion
The read pipe in :mod:`asyncio` UNIX transports now has access to the
`is_reading()` API.
```

**Comment by asvetlov** [comment]
Thanks for NEWs.
The test still should be rewritten.

**Comment by callumquick** [comment]
Thanks for the writeup @asvetlov.
I will take a look at writing a functional test for this change today (and start a new "test case" for Unix read transport functional tests generally).

**Comment by callumquick** [comment]
Hi @asvetlov, I have made the requested changes; please review again.

I added a new file `Lib/test/test_asyncio/test_unix_pipe.py` which defines a new functional test suite for Unix read transport pipes. The test suite sets up two file handles with a read pipe transport registered on the event loop.

Added a test to the suite for verifying the new behaviour of `is_reading()` for `_UnixReadPipeTransport`.

Thanks,
Callum

**Comment by callumquick** [comment]
@asvetlov, please could you look at the changes requested?

**Comment by callumquick** [comment]
@asvetlov, please could you take a look at these updates? Thanks

**Comment by zware** [comment]
@callumquick, after a quick look at the changes here, I think this could do with a clean rebase on master in a new PR (I would suggest just force-pushing over this, but apparently GitHub behaves weirdly with force-push in some situations).

**Comment by callumquick** [comment]
@asvetlov, @zware: New PR with updated summary at  #17755.

**Pull request closed without merging**